### PR TITLE
feature: Support Flat configs in v8

### PR DESCRIPTION
**Pull request by lotmek**
## Description
- Handle flat config in `rule-finder.js`
- Add a new `--flatConfig` flag
- Load ESLint class using FlatESLint class
- Add unit tests for basic flat config

### CODE REVIEW

File: src/bin/find.js [hunk 1]
```diff
@@ -42,7 +45,7 @@ const finderOptions = {
   omitCore: !argv.core,
   includeDeprecated: argv.include === 'deprecated',
   ext: argv.ext,
-  useFlatConfig: argv.flatConfig
+  useFlatConfig: semver.satisfies(eslintPkg.version, '>= 8.57') && argv.flatConfig
```
Comment (ljharb) [inline src/bin/find.js:48]: all this PR seems to do is prevent `useFlatConfig` from being `true` in an eslint without flat config support

**Review by ljharb — CHANGES_REQUESTED** [review]

File: test/lib/rule-finder.js [hunk 2]
```diff
@@ -109,17 +120,28 @@ const getRuleFinder = proxyquire('../../src/lib/rule-finder', {
...
+        return mockedDedupedBuiltinRules
       }
-    }
+    },
+  },
+  "eslint/use-at-your-own-risk": {
+    builtinRules: mockedDedupedBuiltinRules
+  },
+  "../rules": {
+    get: function(id) {
```
Comment (ljharb) [inline test/lib/rule-finder.js:141]:
```suggestion
    get(id) {
```

File: src/bin/find.js [hunk 3]
```diff
@@ -39,7 +44,8 @@ const specifiedFile = argv._[0];
 const finderOptions = {
   omitCore: !argv.core,
   includeDeprecated: argv.include === 'deprecated',
-  ext: argv.ext
+  ext: argv.ext,
+  useFlatConfig: semver.satisfies(eslintPkg.version, '>= 8.57') && argv.flatConfig
```
[the remark this answers is not in the document]
Comment (lotmek) [reply src/bin/find.js]: done 
image

File: src/lib/rule-finder.js [hunk 4]
```diff
@@ -1,11 +1,29 @@
...
+try {
+  builtinRules = require('eslint/use-at-your-own-risk').builtinRules;
+} catch (e) {}
+
+async function _loadEslint(options, useFlatConfig) {
+  if (!useFlatConfig) {
+    // Ignore any config applicable depending on the location on the filesystem
+    options.useEslintrc = false;
+  } else if (!loadESLint) {
+    throw ('This version of ESLint does not support flat config.')
```
Comment (ljharb) [inline src/lib/rule-finder.js:19]:
```suggestion
    throw 'This version of ESLint does not support flat config.';
```

File: src/lib/rule-finder.js [hunk 5]
```diff
@@ -52,26 +72,36 @@ function _notDeprecated(rule) {
...
-    plugins.forEach(plugin => {
-      const normalized = normalizePluginName(plugin);
-      const pluginConfig = require(normalized.module);
-      const rules = pluginConfig.rules === undefined ? {} : pluginConfig.rules;
-
-      Object.keys(rules).forEach(ruleName =>
-        pluginRules.set(`${normalized.prefix}/${ruleName}`, rules[ruleName])
-      );
-    });
+    if(useFlatConfig) {
```
Comment (ljharb) [inline src/lib/rule-finder.js:80]:
```suggestion
    if (useFlatConfig) {
```

File: src/lib/rule-finder.js [hunk 6]
```diff
@@ -52,26 +72,36 @@ function _notDeprecated(rule) {
...
+      Object.entries(config.plugins).forEach(([pluginName, pluginConfig]) => {
+        const rules = pluginConfig.rules === undefined ? {} : pluginConfig.rules;
+  
+        Object.keys(rules).forEach(ruleName =>
+          pluginRules.set(`${pluginName}/${ruleName}`, rules[ruleName])
+        );
+      });
```
Comment (ljharb) [inline src/lib/rule-finder.js:87]:
```suggestion
      Object.entries(config.plugins)
        .filter(([, { rules }) => rules)
        .forEach(([pluginName, { rules }]) => {
          Object.keys(rules).forEach(ruleName =>
            pluginRules.set(`${pluginName}/${ruleName}`, rules[ruleName])
          );
        });
```

File: src/lib/rule-finder.js [hunk 7]
```diff
@@ -52,26 +72,36 @@ function _notDeprecated(rule) {
...
+    } else {
+      plugins.forEach(plugin => {
+        const normalized = normalizePluginName(plugin);
+        const pluginConfig = require(normalized.module);
+        const rules = pluginConfig.rules === undefined ? {} : pluginConfig.rules;
+  
+        Object.keys(rules).forEach(ruleName =>
+          pluginRules.set(`${normalized.prefix}/${ruleName}`, rules[ruleName])
+        );
```
Comment (ljharb) [inline src/lib/rule-finder.js:96]:
```suggestion
        if (pluginConfig.rules) {
          Object.keys(pluginConfig.rules).forEach(ruleName =>
            pluginRules.set(`${normalized.prefix}/${ruleName}`, rules[ruleName])
          );
        }
```

File: src/lib/rule-finder.js [hunk 8]
```diff
@@ -52,26 +72,36 @@ function _notDeprecated(rule) {
...
+      });
+    }
   }
   return pluginRules;
 }
 
 function _getCoreRules() {
-  return new Linter().getRules();
+  return builtinRules ? builtinRules : new Linter().getRules();
```
Comment (ljharb) [inline src/lib/rule-finder.js:104]:
```suggestion
  return builtinRules || new Linter().getRules();
```

**Review by ljharb — CHANGES_REQUESTED** [review]

**Review by ljharb — APPROVED** [review]
LGTM, this will be semver-minor

**Comment by ljharb** [comment]
@lotmek it looks like the exception is always being thrown.

**Comment by lotmek** [comment]
> @lotmek it looks like the exception is always being thrown.

Hello @ljharb, tell me if I'm wrong, but I think flat configs are not supported in ESLint v7

I can add a condition to skip the flat config tests if ESLint version is lower than 8, what do you think of it ?

```js
  if (semver.satisfies(eslintPkg.version, '>= 8')) {
    describe('flat config', () => {
        ...
    })
  }
```

**Comment by ljharb** [comment]
I have eslint 8.56 installed and it's still failing locally, fwiw.

Certainly the flat config tests should be ensuring it throws in eslint 7, good catch.

**Comment by lotmek** [comment]
> I have eslint 8.56 installed and it's still failing locally, fwiw.
> 
> Certainly the flat config tests should be ensuring it throws in eslint 7, good catch.

I did some changes, can you try again ?

**Comment by ljharb** [comment]
LGTM, thanks

**Pull request merged**